zcash_client_sqlite/src/lib.rs

```diff
@@ -230,17 +216,17 @@ impl<C: Borrow<rusqlite::Connection>, P: consensus::Parameters> WalletRead for W
         wallet::get_balance_at(self.conn.borrow(), account, anchor_height)
     }
 
-    fn get_memo(&self, id_note: Self::NoteRef) -> Result<Option<Memo>, Self::Error> {
-        match id_note {
-            NoteId::SentNoteId(id_note) => wallet::get_sent_memo(self.conn.borrow(), id_note),
-            NoteId::ReceivedNoteId(id_note) => {
-                wallet::get_received_memo(self.conn.borrow(), id_note)
-            }
+    fn get_memo(&self, note_id: NoteId) -> Result<Option<Memo>, Self::Error> {
+        let sent_memo = wallet::get_sent_memo(self.conn.borrow(), note_id)?;
+        if sent_memo.is_some() {
+            Ok(sent_memo)
+        } else {
+            wallet::get_received_memo(self.conn.borrow(), note_id)
         }
```

Comment:
This method is now concretely less efficient than before. I don't know by how much, but it is now `O(num_received + num_sent + 2*num_internal)`, because we have two separate table joins against `transactions` (instead of a direct query of an indexed column), and for received notes we always query the sent notes table first. We'll know more once we get this into the mobile SDKs, so this is non-blocking, but we should track its effect on the history-rendering app views.

More non-blocking thoughts: Could we gain back or amortize some of the performance loss by combining the new queries into a single query against all three tables at once? AFAICT this is the only place that `wallet::get_sent_memo` and `wallet::get_received_memo` are used (they were removed from the public API in 0.7.0), so we can just have a single backing function. This does become slightly more annoying given the `sapling_received_notes` vs `sent_notes` split, but we can put off dealing with that to the Orchard work.